src/server/pkg/migrations/migrations.go

```diff
@@ -0,0 +1,172 @@
+package migrations
+
+import (
+	"context"
+	"database/sql"
+	"strings"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
+	"github.com/sirupsen/logrus"
+)
+
+type Env struct {
+	// TODO: etcd
+	ObjectClient obj.Client
+	Tx           *sqlx.Tx
+}
+
+func MakeEnv(objC obj.Client) Env {
+	return Env{
+		ObjectClient: objC,
+	}
+}
+
+type Func func(ctx context.Context, env Env) error
+
+type State struct {
+	n      int
+	prev   *State
+	change Func
+	name   string
+}
+
+func (s State) Apply(name string, fn Func) State {
+	return State{
+		prev:   &s,
+		change: fn,
+		n:      s.n + 1,
+		name:   strings.ToLower(name),
+	}
+}
+
+func (s State) Name() string {
+	return s.name
+}
+
+func (s State) Number() int {
+	return s.n
+}
+
+func InitialState() State {
+	return State{
+		name: "init",
+		change: func(ctx context.Context, env Env) error {
+			_, err := env.Tx.ExecContext(ctx, `CREATE TABLE IF NOT EXISTS migrations (
+				id BIGINT PRIMARY KEY,
+				NAME VARCHAR(250) NOT NULL,
+				start_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+				end_time TIMESTAMP
+			);
+			INSERT INTO migrations (id, name, start_time, end_time) VALUES (0, 'init', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP) ON CONFLICT DO NOTHING;
+			`)
+			return err
+		},
+	}
+}
+
+func ApplyMigrations(ctx context.Context, db *sqlx.DB, baseEnv Env, state State) error {
+	if state.prev != nil {
```

Comment:
We discussed this offline, add a wrapper to convert the linked list to a slice that we iterate through in the code below. This would improve the readability.